### PR TITLE
Setup CQRS for order micro-service

### LINKED ISSUE #74: Setup CQRS for Order Service
- [ ] Create commands
- [ ] Create queries

### PR DESCRIPTION
**Pull request by JortWillemsen**
## Changes
Implemented CQRS in the order micro-service.

## Checklist
- [x] Add commands
- [x] Add  queries
- [x] Add test exclusions
- [x] Add REST controller

## When merging, this pull request 
closes #74

### CODE REVIEW

File: order/src/main/java/nl/softwarestrijders/waiter/order/OrderApplication.java [hunk 1]
```diff
@@ -4,10 +4,15 @@
...
+import java.util.UUID;
+
 @SpringBootApplication
 @TestExcludeGenerated
 public class OrderApplication {
     public static void main(String[] args) {
         SpringApplication.run(OrderApplication.class, args);
+
+        System.out.println("CUSTOMER ID: " + UUID.randomUUID());
+
```
Comment (Rubenvdbrink) [inline order/src/main/java/nl/softwarestrijders/waiter/order/OrderApplication.java:16]: I guess this was for testing purposes? pls remove

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public class OrderController {
+    private final QueryHandler queryHandler;
+    private final CommandHandler commandHandler;
+
+    public OrderController(QueryHandler queryHandler, CommandHandler commandHandler) {
+        this.queryHandler = queryHandler;
+        this.commandHandler = commandHandler;
+    }
+
+    @PostMapping
```
Comment (Rubenvdbrink) [inline order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java:23]: no uri?

**Review by Rubenvdbrink — CHANGES_REQUESTED** [review]
small things

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java (hunk 2 above)
Comment (JortWillemsen) [reply order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java:24]: If you don't specify a uri it just uses the main mapping. In this case /order

File: order/src/main/java/nl/softwarestrijders/waiter/order/OrderApplication.java (hunk 1 above)
Comment (JortWillemsen) [reply order/src/main/java/nl/softwarestrijders/waiter/order/OrderApplication.java]: Yup, removed it

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            @RequestBody ModifyProductOnOrderDto dto) {
+        return this.commandHandler.handleAddProductToOrder(id, dto.productId(), dto.amount());
+    }
+
+    @DeleteMapping("/{id}/product")
+    public void removeProductFromOrder(
+            @PathVariable UUID id,
+            @RequestBody ModifyProductOnOrderDto dto) {
+        this.commandHandler.handleRemoveProductFromOrder(id, dto.productId(), dto.amount());
+    }
```
Comment (xandervedder) [inline order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java:41]:
```suggestion
    public void removeProductFromOrder(
            @PathVariable UUID id,
            @RequestBody ModifyProductOnOrderDto dto
    ) {
        this.commandHandler.handleRemoveProductFromOrder(id, dto.productId(), dto.amount());
    }
```

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    @GetMapping("/{id}")
+    public Order findOrderById(@PathVariable UUID id) {
+        return this.queryHandler.getOrderById(id);
+    }
+
+    @GetMapping
+    public List<Order> findAllOrders(
+            @RequestParam(value = "customer", required = false) UUID customer,
+            @RequestParam(value = "product", required = false) UUID product) {
+
```
Comment (xandervedder) [inline order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java:57]:
```suggestion
    public List<Order> findAllOrders(
            @RequestParam(value = "customer", required = false) UUID customer,
            @RequestParam(value = "product", required = false) UUID product
    ) {

```

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @PostMapping("/{id}/product")
+    public Order addProductToOrder(
+            @PathVariable UUID id,
+            @RequestBody ModifyProductOnOrderDto dto) {
+        return this.commandHandler.handleAddProductToOrder(id, dto.productId(), dto.amount());
+    }
+
+    @DeleteMapping("/{id}/product")
```
Comment (xandervedder) [inline order/src/main/java/nl/softwarestrijders/waiter/order/adapters/http/OrderController.java:36]: nit (maybe this is better, idk):
```suggestion
    @DeleteMapping("/{id}/product/{productId}")
```

File: order/src/main/java/nl/softwarestrijders/waiter/order/core/domain/events/CreatedOrder.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package nl.softwarestrijders.waiter.order.core.domain.events;
+
+import java.util.UUID;
+
+public class CreatedOrder extends OrderEvent {
```
Comment (xandervedder) [inline order/src/main/java/nl/softwarestrijders/waiter/order/core/domain/events/CreatedOrder.java:5]: nit: 
```suggestion
public class OrderCreated extends OrderEvent {
```

File: order/src/main/java/nl/softwarestrijders/waiter/order/core/domain/events/DeletedOrder.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package nl.softwarestrijders.waiter.order.core.domain.events;
+
+import nl.softwarestrijders.waiter.order.core.domain.Order;
+
+import java.util.UUID;
+
+public class DeletedOrder extends OrderEvent {
```
Comment (xandervedder) [inline order/src/main/java/nl/softwarestrijders/waiter/order/core/domain/events/DeletedOrder.java:7]: nit:
```suggestion
public class OrderDeleted extends OrderEvent {
```

**Review by xandervedder — CHANGES_REQUESTED** [review]

File: order/src/main/java/nl/softwarestrijders/waiter/order/core/application/CommandHandler.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+        order.removeProduct(productId, amount);
+
+        this.repository.save(order);
+
+        this.eventPublisher.publish(new ProductRemovedFromOrder(order.getId(), productId, amount));
+    }
+
+    public void handleDeleteOrder(UUID orderId) {
+        var order = this.repository.findById(orderId).orElseThrow(() -> new OrderNotFoundException("Could not find order with id: " + orderId));
```
Comment (JustMilan) [inline order/src/main/java/nl/softwarestrijders/waiter/order/core/application/CommandHandler.java:58]: nit: You could make an utils method for finding it by id

**Review by JustMilan — APPROVED** [review]

File: order/src/main/java/nl/softwarestrijders/waiter/order/core/application/CommandHandler.java (hunk 8 above)
Comment (xandervedder) [reply order/src/main/java/nl/softwarestrijders/waiter/order/core/application/CommandHandler.java:58]: :+1:

File: order/src/main/java/nl/softwarestrijders/waiter/order/adapters/messaging/RabbitMqConfig.java [hunk 9]
```diff
@@ -35,7 +35,7 @@ public Queue orderQueue() {
     }
 
     @Bean
-    public Binding reviewBinding() {
+    public Binding orderBinding() {
```
[the remark this answers is not in the document]
Comment (JortWillemsen) [reply order/src/main/java/nl/softwarestrijders/waiter/order/adapters/messaging/RabbitMqConfig.java:38]: sssht

**Review by Rubenvdbrink — APPROVED** [review]
lgtm

**Review by xandervedder — APPROVED** [review]

**Pull request merged**